aim/agent/aid/universes/aci/tenant.py

```diff
@@ -158,6 +167,139 @@ def _get_instance_subscription_urls(self,
             return urls
 
 
+class OwnershipManager(object):
+
+    def __init__(self, apic_session, apic_config, aim_system_id):
+        self.use_annotation = supports_annotations(apic_session)
+        self.aim_system_id = aim_system_id or apic_config.get_option(
+            'aim_system_id', 'aim')
+        if self.use_annotation:
+            self.annotation_key = apic_config.get_option(
+                'annotation_prefix', 'aim') + self.aim_system_id
+        self.tag_key = self.aim_system_id
+        self.owned_by_tag = set()
+        self.owned_by_annotation = set()
+        self.failure_log = {}
+
+    @property
+    def owned_set(self):
+        return self.owned_by_tag | self.owned_by_annotation
+
+    def set_ownership_key(self, to_push):
+        """Set ownership key
+
+        When pushing objects to APIC, push annotations when supported, TAG
+        only otherwise.
+        :param to_push:
+        :return:
+        """
+        result = to_push
+        if self.use_annotation:
+            for obj in to_push:
+                obj.values()[0]['attributes']['annotation'] = (
+                    self.annotation_key)
+        else:
+            tags = []
+            for obj in to_push:
+                if not obj.keys()[0].startswith(TAG_KEY):
+                    dn = obj.values()[0]['attributes']['dn']
+                    dn += '/tag-%s' % self.tag_key
+                    tags.append({"%s__%s" % (TAG_KEY, obj.keys()[0]):
+                                 {"attributes": {"dn": dn}}})
+            result.extend(tags)
+        return result
+
+    def set_ownership_change(self, to_push):
+        to_update = []
+        to_delete = []
+        if self.use_annotation:
+            for obj in to_push:
+                if obj.keys()[0].startswith(TAG_KEY):
+                    dn = obj.values()[0]['attributes']['dn']
+                    if dn.endswith('/tag-%s' % self.tag_key):
+                        dec = apic_client.DNManager().aci_decompose_dn_guess(
+                            dn, obj.keys()[0])
+                        parent_dec = dec[1][:-1]
+                        parent_dn = apic_client.DNManager().build(parent_dec)
+                        parent_type = parent_dec[-1][0]
+                        to_update.append(
+                            {parent_type: {'attributes': {'dn': parent_dn,
+                                                          'annotation': ''}}})
+                else:
```

Comment:
I think we should get rid of this "else:" and always do "to_delete.append(obj)", otherwise those tags won't get deleted......

Reply:
Good point - will remove. And in fact this goes back to your earlier recommendation to get rid of to_delete all-together - it's the same as to_push, so I'll just use tat.